### ISSUE TITLE
New Rule: `use-public-class-accessor`

### ISSUE THREAD

**Issue by mbullington** [eslint/eslint#15673]
### Rule details

If a get and/or set pair is defined for a private field, warn the user to use it instead of the private field directly.

### Related ECMAScript feature

Private class members

### What type of rule is this?

Warns about a potential problem

### Example code

```js
class A {
    #input: HTMLInputElement;

    get input() {
        // error free since it's in the accessor.
        return this.#input;
    }

    set input(val) {
        // error free since it's in the accessor.
        this.#input = val;

        // side-effects here that would be missed if you
        // just did `this.#input = abc`!
        jobQueue.push(val)
    }

    someOtherFunction() {
        // should cause 'use-public-class-accessor' 
        this.#input = document.createElement('input');

        // should cause 'use-public-class-accessor' 
        console.log(this.#input);
    }
}
```

### Why should this rule be in the core instead of a plugin?

This seems generic enough that to create a separate plugin for this, it would just be `eslint-plugin-use-public-class-accessor`.

I believe for imperative workflows (such as UI with web components) this is incredibly useful, while not directly tying into any of those concepts.

### Participation

- [X] I am willing to submit a pull request to implement this rule.

### Additional comments

Please let me know if there's already an implementation of a similar pattern I've just missed, or an existing rule that could be extended to cover this use-case.

(also if the name should be changed.)

Best,

**Comment by nzakas** [comment]
Thanks for the suggestion. I’m not sure that this pattern is actually a problem, or that we can logically infer that public methods shouldn’t use private fields. It seems more like a stylistic preference to me, and in that case, I don’t think it’s appropriate for the core. Let’s see what others think.

**Comment by ljharb** [comment]
I agree; private fields are often - if not primarily - used in public methods, and the Airbnb style guide (to name a popular example) strictly prohibits getters and setters entirely.

**Comment by mdjermanovic** [comment]
I agree with @nzakas. This looks more like a project-specific convention than a general rule and is therefore more appropriate for a custom rule than for a core rule.

**Comment by nzakas** [comment]
Thanks for the submission. Unfortunately, this rule doesn't meet our [criteria for new rules](https://eslint.org/blog/2020/05/changes-to-rules-policies), so I'm closing this pull request. Please note that you can always create any rule you want and distribute it in your own plugin for use with ESLint.
